File: SADCOMS.Domain/Entities/Order.cs
using System.ComponentModel.DataAnnotations;
using SADCOMS.Domain.Enums;

namespace SADCOMS.Domain.Entities;

public class Order : BaseEntity
{
  public Guid CustomerId { get; set; }
  public Customer Customer { get; set; }
  public OrderStatus Status { get; set; }
  public required string CurrencyCode { get; set; }
  public decimal TotalAmount { get; set; }
  [Timestamp]
  public byte[] RowVersion { get; set; }
  public ICollection<OrderLineItem> LineItems { get; set; } = new List<OrderLineItem>();

  public void RecalculateTotalAmount()
  {
    TotalAmount = LineItems.Sum(li => li.Quantity * li.UnitPrice);
  }

  private static readonly Dictionary<OrderStatus, OrderStatus[]> _validTransitions = new()
{
    { OrderStatus.Pending,   new[] { OrderStatus.Paid, OrderStatus.Cancelled } },
    { OrderStatus.Paid,      new[] { OrderStatus.Fulfilled, OrderStatus.Cancelled } },
    { OrderStatus.Fulfilled, Array.Empty<OrderStatus>() },
    { OrderStatus.Cancelled, Array.Empty<OrderStatus>() },
};

  public void UpdateStatus(OrderStatus newStatus)
  {
    var allowed = _validTransitions[Status];
    if (!allowed.Contains(newStatus))
    {
      throw new InvalidOperationException(
          $"Cannot transition from {Status} to {newStatus}.");
    }
    Status = newStatus;
  }
}

File: SADCOMS.Web/src/types/Order.ts
export interface Order{
  id: string;
  customerId: string;
  status: OrderStatus;
  currencyCode: string;
  totalAmount: number;
  createdAt: string;
  lineItems: Array<OrderLineItem>;
}

// Can't use enum here — erasableSyntaxOnly requires all TypeScript syntax to be erasable, and enums emit runtime JS
export const OrderStatus = {
  Pending: 0,
  Paid: 1,
  Fulfilled: 2,
  Cancelled: 3,
} as const;

export type OrderStatus = typeof OrderStatus[keyof typeof OrderStatus];


export interface OrderLineItem{
  id: string;
  productSku: string;
  quantity: number;
  unitPrice: number;
}
